apm-agent-core/src/main/java/co/elastic/apm/agent/impl/CloudMetadataProvider.java

```diff
@@ -0,0 +1,401 @@
+     */
+    @Nullable
+    static CloudProviderInfo fetchAndParseCloudProviderInfo(final CoreConfiguration.CloudProvider cloudProvider, final int queryTimeoutMs) {
+
+        Throwable unexpectedError = null;
+        CloudProviderInfo cloudProviderInfo = null;
+        try {
+            switch (cloudProvider) {
+                case AWS: {
+                    try {
+                        cloudProviderInfo = getAwsMetadata(queryTimeoutMs, cloudProvider);
+                    } catch (Exception e) {
+                        unexpectedError = e;
+                    }
+                    break;
+                }
+                case GCP: {
+                    try {
+                        cloudProviderInfo = getGcpMetadata(queryTimeoutMs);
+                    } catch (Exception e) {
+                        unexpectedError = e;
+                    }
+                    break;
+                }
+                case AZURE: {
+                    try {
+                        cloudProviderInfo = getAzureMetadata(queryTimeoutMs);
+                    } catch (Exception e) {
+                        unexpectedError = e;
+                    }
+                    break;
+                }
+                case AUTO: {
+                    ExecutorService executor = ExecutorUtils.createThreadDaemonPool("apm-cloud-metadata", 2, 2);
+                    Future<CloudProviderInfo> awsMetadata;
+                    Future<CloudProviderInfo> gcpMetadata;
+                    try {
+                        awsMetadata = executor.submit(new Callable<CloudProviderInfo>() {
+                            @Override
+                            public CloudProviderInfo call() {
+                                CloudProviderInfo awsInfo = null;
+                                try {
+                                    awsInfo = getAwsMetadata(queryTimeoutMs, cloudProvider);
+                                } catch (Exception e) {
+                                    // Expected - trial and error method
+                                }
+                                return awsInfo;
+                            }
+                        });
+
+                        gcpMetadata = executor.submit(new Callable<CloudProviderInfo>() {
+                            @Override
+                            public CloudProviderInfo call() {
+                                CloudProviderInfo gcpInfo = null;
+                                try {
+                                    gcpInfo = getGcpMetadata(queryTimeoutMs);
+                                } catch (Exception e) {
+                                    // Expected - trial and error method
+                                }
+                                return gcpInfo;
+                            }
+                        });
+                    } finally {
+                        executor.shutdown();
+                    }
+
+                    long futureTimeout = queryTimeoutMs + 200;
```

Comment:
[question] why do we add a fixed extra timeout for the future, how do we know that 200ms is fine ?

Reply:
> why do we add a fixed extra timeout for the future

We want to limit this blocking somehow, but if I use the same timeout used for the HTTP query, that would be too tight, it needs to be longer.

> how do we know that 200ms is fine ?

We don't know, we live on the edge 😛